tools/isobuild/import-scanner.js

```diff
@@ -0,0 +1,431 @@
+      info.path = pathJoin("app", info.path);
+    }
+
+    // Note that info.inNodeModules may be false even if info.path now
+    // contains a node_modules directory.
+    return info;
+  }
+
+  _getNodeModulesInstallInfo(absPath) {
+    if (this.nodeModulesPath) {
+      const relPathWithinNodeModules =
+        pathRelative(this.nodeModulesPath, absPath);
+
+      if (relPathWithinNodeModules.startsWith("..")) {
+        // absPath is not a subdirectory of this.nodeModulesPath.
+        return;
+      }
+
+      if (! this._hasKnownExtension(relPathWithinNodeModules)) {
+        // Only accept files within node_modules directories if they
+        // have one of the known extensions.
+        return;
+      }
+
+      return {
+        // Install the module into the local node_modules directory within
+        // this app or package.
+        path: pathJoin("node_modules", relPathWithinNodeModules),
+        // The original path was contained by a node_modules directory.
+        inNodeModules: true,
+      };
+    }
+  }
+
+  _getSourceRootInstallInfo(absPath) {
+    const installPath = pathRelative(this.sourceRoot, absPath);
+
+    if (installPath.startsWith("..")) {
+      // absPath is not a subdirectory of this.sourceRoot.
+      return;
+    }
+
+    const dirs = this._splitPath(pathDirname(installPath));
+    const bundlingClientApp =
+      ! this.name && // Indicates we are bundling an app.
+      archMatches(this.bundleArch, "web");
+
+    for (let dir of dirs) {
+      if (dir.charAt(0) === "." ||
+          dir === "packages" ||
+          dir === "programs" ||
+          dir === "cordova-build-override") {
+        // These directories are never loaded as part of an app.
+        return;
+      }
+
+      if (bundlingClientApp && (dir === "server" ||
+                                dir === "private")) {
+        // If we're bundling an app for a client architecture, any files
+        // contained by a server-only directory that is not contained by
+        // a node_modules directory must be ignored.
+        return;
+      }
+
+      if (dir === "node_modules") {
+        if (! this._hasKnownExtension(installPath)) {
+          // Reject any files within node_modules directories that do
+          // not have one of the known extensions.
+          return;
+        }
+
+        // Accept any file within a node_modules directory if it has a
+        // known file extension.
+        return {
+          path: installPath,
+          inNodeModules: true,
+        };
+      }
+    }
+
+    return {
+      path: installPath,
+      inNodeModules: false,
+    };
+  }
+
+  _hasKnownExtension(path) {
+    return this.extensions.indexOf(pathExtname(path)) >= 0;
+  }
+
+  _splitPath(path) {
+    const partsInReverse = [];
+    for (let dir; (dir = pathDirname(path)) !== path; path = dir) {
+      partsInReverse.push(pathBasename(path));
+    }
+    return partsInReverse.reverse();
+  }
+
+  _tryToResolveImportedPath(id, path) {
```

Comment:
I think this function, and honestly almost every function in this class, should have a comment.